crates/gazette/src/journal/append.rs

```diff
@@ -0,0 +1,186 @@
+                }
+            }
+        })
+    }
+
+    async fn append_all<R>(
+        &self,
+        journal: &str,
+        source: &mut R,
+    ) -> crate::Result<broker::AppendResponse>
+    where
+        R: AsyncBufReadExt + Send + Unpin,
+    {
+        // Transforms `source` into a stream of `Result<AppendRequest, gazette::Error>`. This deals with
+        // the append RPC's semantics that require an initial "metadata" request, followed by a stream of
+        // "chunk" requests, followed by an empty request to indicate we're done. Potential errors ultimately
+        // originate from reading the input AsyncRead.
+        let request_generator = coroutines::coroutine(move |mut co| async move {
+            // Send initial request
+            () = co
+                .yield_(Ok(broker::AppendRequest {
+                    journal: journal.to_string(),
+                    ..Default::default()
+                }))
+                .await;
+
+            loop {
+                // Process chunks until EOF
+                let bytes_read = match source.fill_buf().await {
+                    // An empty buffer indicates EOF, as otherwise fill_buf() will wait until data is available
+                    Ok(chunk) if chunk.len() == 0 => break,
+                    Ok(chunk) => {
+                        () = co
+                            .yield_(Ok(broker::AppendRequest {
+                                content: chunk.to_vec(),
+                                ..Default::default()
+                            }))
+                            .await;
+                        chunk.len()
+                    }
+                    Err(e) => {
+                        () = co.yield_(Err(Error::Append(e.to_string()))).await;
+                        return;
+                    }
+                };
+
+                source.consume(bytes_read);
+            }
+            // Send final empty chunk
+            () = co
+                .yield_(Ok(broker::AppendRequest {
+                    ..Default::default()
+                }))
+                .await;
+        });
+
+        // Since reading from `source` can error, we need this whole song and dance to
+        // handle those errors. We could just `.collect()` all of the requests and catch
+        // any errors there, but since this is supposed to handle significant volumes of data
+        // over an undefined period of time, that won't work. So instead we need to pass
+        // `JournalClient::append()` a stream of _just_ the `AppendRequest`s that come out
+        // of the above `request_generator`, while also promptly returning any errors if they
+        // crop up, and cancelling the append request.
+
+        let (req_tx, req_rx) = tokio::sync::mpsc::channel(100);
+
+        let mut client = self.into_sub(self.router.route(None, false, &self.default)?);
+
+        // Run `JournalClient::append` in a separate Tokio task, and feed it a steady diet of `AppendRequest`s
```

Comment:
We should avoid using tokio::spawn altogether. I don't believe it's necessary, and it significantly complicates the "drop to cancel" story.

Performing a concurrent poll of the next chunk of bytes AND polling to send an AppendRequest is a good idea, but that can also be achieved using `futures::join` instead.